go-controller/pkg/node/vrfmanager/vrf_manager.go

```diff
@@ -0,0 +1,321 @@
+					}
+				case <-linkSyncTimer.C:
+					if err = vrfm.reconcile(); err != nil {
+						klog.Errorf("VRF Manager: Error while reconciling vrfs, err: %v", err)
+					}
+				case <-stopChan:
+					close(doneCh)
+					return nil
+				}
+				close(doneCh)
+			}
+		})
+		if err != nil {
+			klog.Errorf("VRF Manager: failed to run link reconcile goroutine, err: %v", err)
+		}
+	}()
+	klog.Info("VRF manager is running")
+	return nil
+}
+
+func (vrfm *Controller) reconcile() error {
+	vrfm.mu.Lock()
+	defer vrfm.mu.Unlock()
+	start := time.Now()
+	defer func() {
+		klog.V(5).Infof("VRF Manager: reconciling VRFs took %v", time.Since(start))
+	}()
+
+	for _, vrf := range vrfm.vrfs {
+		if vrf.delete {
+			err := vrfm.deleteVRF(vrf)
+			if err != nil {
+				klog.Errorf("VRF Manager: error deleting VRF device %s during reconcile, err: %v", vrf.name, err)
+			} else {
+				delete(vrfm.vrfs, vrf.name)
+			}
+			continue
+		}
+		err := vrfm.sync(vrf)
+		if err != nil {
+			klog.Errorf("VRF Manager: error syncing VRF device %s during reconcile, err: %v", vrf.name, err)
+		}
+	}
+	return nil
+}
+
+func (vrfm *Controller) syncLink(link netlink.Link) error {
+	vrfm.mu.Lock()
+	defer vrfm.mu.Unlock()
+	vrf, ok := vrfm.vrfs[link.Attrs().Name]
+	if !ok {
+		return nil
+	}
+	return vrfm.sync(vrf)
+}
+
+func (vrfm *Controller) sync(vrf vrf) error {
+	vrfLink, err := util.GetNetLinkOps().LinkByName(vrf.name)
+	var mustRecreate bool
+	if err == nil {
+		if vrfLink.Type() != "vrf" {
+			return fmt.Errorf("node has another non vrf device with same name %s", vrf.name)
+		}
+		vrfDev, ok := vrfLink.(*netlink.Vrf)
+		if ok && vrfDev.Table != vrf.table {
+			klog.Warningf("Found a conflict with existing vrf device table id for vrf device %s, recreating it", vrf.name)
+			err = vrfm.deleteVRF(vrf)
+			if err != nil {
+				return fmt.Errorf("failed to delete existing vrf device %s to recreate, err: %w", vrf.name, err)
+			}
+			mustRecreate = true
+		}
+	}
+	// Create VRF device if it doesn't exist or if it's needed to be recreated.
+	if util.GetNetLinkOps().IsLinkNotFoundError(err) || mustRecreate {
+		vrfLink = &netlink.Vrf{
+			LinkAttrs: netlink.LinkAttrs{Name: vrf.name},
+			Table:     vrf.table,
+		}
+		if err = util.GetNetLinkOps().LinkAdd(vrfLink); err != nil {
+			return fmt.Errorf("failed to create vrf device %s, err: %v", vrf.name, err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("failed to retrieve existing vrf device %s, err: %v", vrf.name, err)
+	}
+	vrfLink, err = util.GetNetLinkOps().LinkByName(vrf.name)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve vrf device %s, err: %v", vrf.name, err)
+	}
+	if vrfLink.Attrs().OperState != netlink.OperUp {
+		if err = util.GetNetLinkOps().LinkSetUp(vrfLink); err != nil {
+			return fmt.Errorf("failed to get vrf device %s operationally up, err: %v", vrf.name, err)
+		}
+	}
+	existingEnslave, err := getExistingEnslaveInterfaceForVRF(vrfLink)
```

Comment:
this function searches all links and just finds the first one. We shouldn't assume there is only one. i think we should get them all